### PR TITLE
closed #3874: expand textureID bit to 32bits

### PR DESCRIPTION
**Pull request by dabingnn**

### CODE REVIEW

File: cocos/2d/renderer/CCQuadCommand.cpp [hunk 1]
```diff
@@ -92,13 +92,13 @@ void QuadCommand::generateMaterialID()
...
-    // +---------------------+-------------------+----------------------+
+    // +---------------------+-------------------+----------------------------------------+
+    // | Shader ID (10 bits) | Blend ID (4 bits) | empty (18bits) |  Texture ID (32 bits) |
+    // +---------------------+-------------------+----------------------------------------+
 
-    _materialID = (uint32_t)_shader->getProgram() << 22
-            | (uint32_t)blendID << 18
-            | (uint32_t)_textureID << 0;
+    _materialID = (uint64_t)_shader->getProgram() << 54
+            | (uint64_t)blendID << 40
```
Comment (robertjerovsek) [inline cocos/2d/renderer/CCQuadCommand.cpp]: Shouldn't the blendID be shifted by 50 bits?

Comment (dabingnn) [inline cocos/2d/renderer/CCQuadCommand.cpp]: Thank you, I will fix it.

**Comment by dumganhar** [comment]
Thanks @dabingnn . :)

**Comment by ricardoquesada** [comment]
thanks.
Was there any performance penalty after applying this patch ?

I would guess that no. But it is always good to do a performance test on an old 32-bit device to make sure that the performance it is still good.

**Comment by dabingnn** [comment]
I profiled this patch, the method involved in materialID consumption are generateMaterialID and getMaterialID, they only consume 0.1-0.2% of the total CPU. Besides, there are no FPS decrease on normal tests and performance tests(tests on spritePerformance test, 3000 sprites). Given this results, I think there is no performance problem when migrating materialID to 64bits.

**Pull request merged**